cors.go

```diff
@@ -1,22 +1,113 @@
 package handlers
 
-import "net/http"
+import (
+	// "log"
+	"net/http"
+	"strconv"
+	"strings"
+)
 
 // CORSOption represents a functional option for configuring the CORS middleware.
 type CORSOption func(*cors) error
 
 type cors struct {
-	h              http.Handler
-	allowedHeaders []string
-	allowedMethods []string
-	allowedOrigins []string
-	maxAge         int
-	ignoreOptions  bool
+	h                http.Handler
+	allowedHeaders   []string
+	allowedMethods   []string
+	allowedOrigins   []string
+	exposedHeaders   []string
+	maxAge           int
+	ignoreOptions    bool
+	allowCredentials bool
 }
 
+var (
+	defaultCorsMethods = []string{"GET", "HEAD", "POST"}
+	defaultCorsHeaders = []string{"Accept", "Accept-Language", "Content-Language"}
+)
+
+const (
+	corsOptionMethod           string = "OPTIONS"
+	corsAllowOriginHeader      string = "Access-Control-Allow-Origin"
+	corsExposeHeadersHeader    string = "Access-Control-Expose-Headers"
+	corsMaxAgeHeader           string = "Access-Control-Max-Age"
+	corsAllowMethodsHeader     string = "Access-Control-Allow-Methods"
+	corsAllowHeadersHeader     string = "Access-Control-Allow-Headers"
+	corsAllowCredentialsHeader string = "Access-Control-Allow-Credentials"
+	corsRequestMethodHeader    string = "Access-Control-Request-Method"
+	corsRequestHeadersHeader   string = "Access-Control-Request-Headers"
+	corsOriginHeader           string = "Origin"
+	corsVaryHeader             string = "Vary"
+	corsOriginMatchAll         string = "*"
+)
+
 func (ch *cors) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	handler := ch.h
+	defer func() {
+		handler.ServeHTTP(w, r)
+	}()
+
+	if origin := r.Header.Get(corsOriginHeader); ch.isOriginAllowed(origin) {
+		if r.Method == corsOptionMethod {
+			if ch.ignoreOptions {
+				return
+			}
+
+			handler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) { return })
```

Comment:
I'm not sure of the implications of creating a new HandlerFunc here every time this is called. I think it should be declared as a global and just assigned here otherwise it may be creating unnecessary garbage.